### PR TITLE
Catch calender events fetch errors and keep UI working

### PR DESCRIPTION
**Pull request by spacegaier**
## Proposed change

I have a few old calendar entities in my instance that are no longer returning events through the API call to `/api/calendars` that the frontend triggers. Instead they are resulting in a thrown HTTP 400 which meant we got an runtime exception which in return stopped the UI from working. E.g. the next/prev buttons to switch the month/week/day stopped (the displayed month/week/day did not change).

Not sure if there is a better / more elegant way to handle that situation though...

## Type of change

- [ ] Dependency upgrade
- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [X] The code change is tested and works locally.
- [X] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Comment by allenporter** [comment]
I don't have a strong opinion on the frontend side of this, but i'm wondering if these should not be returned from core in the first place? I would expect any calendar returned from the calendar list shouldn't 404.  I would be happy to update core to fix this if you can tell me a bit more about the state of the invalid calendars. I assume we want them not to show up on the list on the left in that case.

**Comment by spacegaier** [comment]
Unfortunately, I can't provide more details anymore, since they are now working again (I think all non local calendars in this HA instance are Google calendars that apparently had temporarily an issue and therefore returned the HTTP error).

I agree that ideally those scenarios would be handled by the backend, but if we cannot rule them out, then I think the frontend should at least stay responsive if data could not be read (and perhaps also show toast to inform the user), so I think this PR still makes sense.

**Comment by allenporter** [comment]
Makes sense for a transient error -- I was thinking it was a permanent error that was ultimately a server bug.

File: src/data/calendar.ts [hunk 1]
```diff
@@ -71,7 +71,15 @@ export const fetchCalendarEvents = async (
     );
   });
 
-  const results = await Promise.all(promises);
+  let results;
+  try {
+    results = await Promise.all(promises);
```
Comment (bramkragten) [inline src/data/calendar.ts:76]: Now if one of the promises fail, everything fails. We should use `allSettled` so we can still show the working ones.
Also we should show an error message for the failing ones.

```suggestion
    results = await Promise.allSettled(promises);
```

Comment (spacegaier) [reply src/data/calendar.ts:76]: `allSettled()` requires a newer JS profile or a shim.

Where do you envision the error message, e.g. if the calendar is shown in a calendar card in a dashboard where one of two configured calendars failed?

Comment (bramkragten) [reply src/data/calendar.ts:76]: https://github.com/zloirock/core-js/blob/master/packages/core-js/modules/es.promise.all-settled.js

The error could be above the `ha-calendar` element and dismissible.

Comment (bramkragten) [reply src/data/calendar.ts:76]: Or just don't use `await Promise.all` :-) but await in the step below

Comment (spacegaier) [reply src/data/calendar.ts:76]: > Or just don't use `await Promise.all` :-) but await in the step below

Not sure what you mean by that TBH. If we were to use `allSettled` would we then add the linked module as a real dep or just copy that coding in with a reference to the repo?

Comment (spacegaier) [reply src/data/calendar.ts:76]: And we also need a way for the caller of `fetchCalendarEvents()` to know about the errors, so just returning the `Promise` might not be sufficient, but instead we also need e.g. a failure counter. I never really took some time to read about this whole async `Promise` topic, so not really my jam yet...

**Comment by spacegaier** [comment]
Done for ha-calendar-card, but not ha-panel-calendar yet.

image

**Review by bramkragten — DISMISSED** [review]

**Review by bramkragten — APPROVED** [review]

**Pull request merged**